org-code-javabuilder/lib/src/main/java/org/code/javabuilder/UserClassLoader.java

```diff
@@ -0,0 +1,95 @@
+package org.code.javabuilder;
+
+import static org.code.protocol.LoggerNames.MAIN_LOGGER;
+
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.logging.Logger;
+import org.json.JSONObject;
+
+/**
+ * Custom class loader for user-provided code. This class loader only allows certain classes to be
+ * used within a user-provided class.
+ */
+public class UserClassLoader extends URLClassLoader {
+  private final List<String> userProvidedClasses;
+  private final URLClassLoader approvedClassLoader;
+
+  public UserClassLoader(URL[] urls, ClassLoader parent, List<String> userProvidedClasses) {
+    super(urls, parent);
+    this.userProvidedClasses = userProvidedClasses;
+    this.approvedClassLoader = new URLClassLoader(urls, JavaRunner.class.getClassLoader());
+  }
+
+  @Override
+  public Class<?> loadClass(String name) throws ClassNotFoundException {
+    // Call super for user provided classes, as we need to verify users are not
+    // trying to use an unapproved class or package.
+    if (this.userProvidedClasses.contains(name)) {
+      return super.loadClass(name);
+    }
+    // If this is not a user provided class, we are loading something used by a user provided class.
+    // If it is either an allowed class or package, we can load with our standard class loader.
+    // Otherwise, throw an exception.
+    if (this.allowedClasses.contains(name)) {
+      return this.approvedClassLoader.loadClass(name);
+    }
+    // allow .<specific-class> usage from valid packages
+    for (int i = 0; i < this.allowedPackages.length; i++) {
+      if (name.startsWith(this.allowedPackages[i])) {
+        return this.approvedClassLoader.loadClass(name);
+      }
+    }
```

Comment:
Does this have any noticeable performance impact? It's a pretty small list so I wouldn't think so, but since this check happens for every non-user class I wasn't sure.

Reply:
Also should we be checking for the `.` explicitly here? It's probably unlikely, but could we end up matching some class in a different package like hypothetically `java.utils.SomeClass` since it starts with `java.util`?

Reply:
I've updated this to have `.` at the end of the approved packages, and I changed the individual class list to a set for faster lookup.